### PR TITLE
Add support for GradientStroke shape items

### PR DESCRIPTION
**Pull request by calda**
This PR adds support for `GradientStroke` shape items, which the Core Animation rendering engine previously silently ignored.

This implementation in the new rendering engine performs quite a lot better than the main thread rendering engine on some examples:

| Main Thread rendering engine | Core Animation rendering engine |
| ------ | ----- |
|  |  |

### CODE REVIEW

File: Sources/Private/Experimental/Animations/ShapeAnimation.swift [hunk 1]
```diff
@@ -42,43 +42,17 @@ extension CAShapeLayer {
...
-  }
-
   /// Adds animations for `strokeStart` and `strokeEnd` from the given `Trim` object
   @nonobjc
   func addAnimations(for trim: Trim, context: LayerAnimationContext) {
     let (strokeStartKeyframes, strokeEndKeyframes) = trim.caShapeLayerKeyframes()
 
+    if trim.offset.keyframes.contains(where: { $0.value.cgFloatValue != 0 }) {
+      LottieLogger.shared.assertionFailure("""
+      The CoreAnimation rendering engine doesn't support Trim offsets
```
Comment (calda) [inline Sources/Private/Experimental/Animations/ShapeAnimation.swift:52]: Some of the examples I was working with use `trim.offset`, which I hadn't encountered before. I investigated a bit and it seems pretty tough to support properly. 

In two examples I was able to work around this by removing the `trim.offset` from the animation json and just adding a corresponding `rotation`. That only works for circles, though, and may not generalize to all types of `offset` values 🤷🏼‍♂️

File: Sources/Private/Experimental/Animations/StrokeAnimation.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+extension GradientStroke: StrokeShapeItem {
+  var strokeColor: KeyframeGroup<Color>? { nil }
+}
+
+// MARK: - CAShapeLayer + StrokeShapeItem
+
+extension CAShapeLayer {
+  /// Adds animations for properties related to the given `Stroke` object (`strokeColor`, `lineWidth`, etc)
+  @nonobjc
+  func addStrokeAnimations(for stroke: StrokeShapeItem, context: LayerAnimationContext) {
```
Comment (calda) [inline Sources/Private/Experimental/Animations/StrokeAnimation.swift:37]: Moved this here from another file

File: Sources/Private/Experimental/Animations/ShapeAnimation.swift [hunk 3]
```diff
@@ -42,43 +42,17 @@ extension CAShapeLayer {
       context: context)
   }
 
-  /// Adds animations for properties related to the given `Stroke` object (`strokeColor`, `lineWidth`, etc)
-  @nonobjc
-  func addAnimations(for stroke: Stroke, context: LayerAnimationContext) {
```
Comment (calda) [inline Sources/Private/Experimental/Animations/ShapeAnimation.swift:47]: Moved this to https://github.com/airbnb/lottie-ios/pull/1516/files#r804256196

File: Sources/Private/Experimental/Layers/GradientRenderLayer.swift [hunk 4]
```diff
@@ -59,7 +59,7 @@ final class GradientRenderLayer: CAGradientLayer {
   ///  - This specific value is arbitrary and can be increased if necessary.
   ///    Theoretically this should be "infinite", to match the behavior of
   ///    `CGContext.drawLinearGradient` with `[.drawsAfterEndLocation, .drawsBeforeStartLocation]`.
-  private let gradientPadding: CGFloat = 1_000
+  private let gradientPadding: CGFloat = 2_000
```
Comment (calda) [inline Sources/Private/Experimental/Layers/GradientRenderLayer.swift:62]: I found a case where this padding was too small, so I bumped it higher.

| Before (the stroke and gradient don't completely overlap) | After |
| ----- | ----- |
| Screen Shot 2022-02-10 at 2 03 07 PM | Screen Shot 2022-02-10 at 2 05 58 PM |

Comment (calda) [reply Sources/Private/Experimental/Layers/GradientRenderLayer.swift:62]: that layer having a bounds of like 10000 is probably why it performs so poorly in the main thread engine 😛

File: Sources/Private/Experimental/Layers/ShapeItemLayer.swift [hunk 5]
```diff
@@ -59,32 +59,51 @@ final class ShapeItemLayer: CALayer {
     let parentGroup: Group?
   }
 
-  override func layoutSublayers() {
-    super.layoutSublayers()
+  override func setupAnimations(context: LayerAnimationContext) {
+    super.setupAnimations(context: context)
 
-    for sublayer in sublayerConfiguration?.layers ?? [] {
-      sublayer.fillBoundsOfSuperlayer()
```
Comment (calda) [inline Sources/Private/Experimental/Layers/ShapeItemLayer.swift:66]: Adding a `BaseAnimationLayer` conformance lets us remove this custom layout code

File: Tests/Samples/LottieFiles/loading_gradient_strokes.json [hunk 6]
```diff
@@ -0,0 +1 @@
...1,"ml":4,"nm":"Gradient Stroke 1","mn":"ADBE Vector Graphic - G-Stroke"},{"ty":"tr","p":{"a":0,"k":[0,0],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transform"}],"nm":"Ellipse 1","np":2,"cix":2,"ix":1,"mn":"ADBE Vector Group"},{"ty":"tm","s":{"a":1,"k":[{"i":{"x":[0.5],"y":[1]},"o":{"x":[0.167],"y":[0]},"n":["0p5_1_0p167_0"],"t":0,"s":[0],"e":[0]},{"i":{"x":[0.5],"y":[1]},"o":{"x":[0.5],"y":[0]},"n":["0p5_1_0p5_0"],"t":10,"s":[0],"e":[99]},{"t":81}],"x":"var $bm_rt;\n$bm_rt = loopOut('Cycle');","ix":1},"e":{"a":1,"k":[{"i":{"x":[0.5],"y":[1]},"o":{"x":[0.5],"y":[0]},"n":["0p5_1_0p5_0"],"t":0,"s":[1],"e":[100]},{"i":{"x":[0.833],"y":[1]},"o":{"x":[0.5],"y":[0]},"n":["0p833_1_0p5_0"],"t":71,"s":[100],"e":[100]},{"t":81}],"x":"var $bm_rt;\n$bm_rt = loopOut('Cycle');","ix":2},"o":{"a":1,"k":[{"i":{"x":[0.5],"y":[1]},"o":{"x":[0.5],"y":[0]},"n":["0p5_1_0p5_0"],"t":0,"s":[0],"e":[360]},{"t":81}],"x":"var $bm_rt;\n$bm_rt = loopOut('Cycle');","ix":3},"m":1,"ix":2,"nm":"Trim Paths 1","mn":"ADBE Vector Filter - Trim"}],"ip":0,"op":180,"st":0,"bm":0,"sr":1}]}
```
Comment (calda) [inline Tests/Samples/LottieFiles/loading_gradient_strokes.json:1]: Added an animation from LottieFiles that uses `GradientStroke` to make sure we have a test for this in the public regression suite

**Review by erichoracek — APPROVED** [review]
LGTM!

**Pull request merged**